### ISSUE TITLE
Style guide structure & formatting

### ISSUE THREAD

**Issue by redoPop** [tedconf/code-style-guides#19]
Until now, this Code Style Guides repo [has been a collection of flat markdown files](https://github.com/tedconf/code-style-guides/tree/2207b126a7074153e56a0aa681f2ffac0a45b17e):

![1-old-busted](https://cloud.githubusercontent.com/assets/109632/11251404/b5646b28-8df6-11e5-99c4-5282a0ad48b1.png)

b3a2875dfe6a19beaf74ee5f3763ac1e11fe7f6d experiments with making it a fully structured Ember app [with its own site](http://tedconf.github.io/code-style-guides/):

![2-new-hotness](https://cloud.githubusercontent.com/assets/109632/11251406/b8777076-8df6-11e5-8e34-eb148e89defd.png)

This change results in the style guide documents themselves using the structure of an Ember app (i.e. [they are stored as templates within pods](https://github.com/tedconf/code-style-guides/blob/40078c70d245a3636f95341752bc8d97ac5251bf/app/pods/ruby/rspec/template.md)).

This issue exists to discuss the pros & cons of this layout and rendering, and decide which of those factors weigh more strongly for us as a team.

**Comment by samselikoff** [comment]
Thanks for kicking off this issue @redoPop!

For those of you who missed our FlowDock conversation, a quick recap. Until yesterday, the `code-style-guides` repo had four Markdown documents in its root, and a fledgling Ember app in `/frontend`. This repo was originally created in June, and has seen little activity since then. The purpose of the repo is for us to have a place to share and canonize how we write code at TED.

The main original content is Joe's CSS document, and an Ember forms document that Ryan, Brenna and I collaborated on. The Ember app was an experiment I worked on several weeks ago, to see what it would be like if the docs were made into a website.

Last night, I worked on that experiment some more. I've always favored Markdown for documentation, since the barriers to entry for editing it are so low. This is what lead me to use Jekyll for Mirage's docs site, even though I would have liked to use Ember (for a variety of reasons, more on that later).

In any case, last night I worked on a Broccoli plugin that would give me the best of both worlds: the ability to author Ember templates as Markdown files. This would allow us to keep our code style guides as Markdown files, but allow us to publish these guides within an Ember app.

This morning, I went ahead and moved the existing Markdown documents within `code-style-guides` into the new app and deployed it to http://tedconf.github.io/code-style-guides/. I did this on my own without input from the team, because (a) sometimes I get excited and just *do*, and (b) since I personally thought the move was an unmitigated improvement, I thought it was not going to be controversial.

Today on Flowdock Joe raised some arguments *against* abandoning the flat Markdown documents. I will let him defend those arguments here, but suffice it to say the move warrants more discussion among our team. Also, I want to be clear that even though I feel strongly about the virtues of having `code-style-guides` be an Ember app, by merging my work directly into `master` and deploying the app this morning I was *not* trying to assert my will over this decision. I simply did not consider the downsides.

Ok, so that's the setup. Here are my arguments in favor of the proposal: *make code-style-guides an Ember app.*

**Why should `code-style-guides` be a website in the first place?**

In my mind a website would provide several benefits over a plain repo:

1. **Ease of consumption**
    - *Navigability* Assuming we start writing more content, a website would give us better control over navigation than a directory of files (think: nav bar or structured ToC vs. lexicographic sorting)
    - *Better reading experience* A bespoke website would let us format our guides so they're front-and-center, without any cruft.
    - Ultimately, the argument here is the same as for why our Tech blog exists: the Middleman files are all Markdown documents, rendered in GitHub, linkable, etc., yet we publish them to a website, since that's a better consumption format and makes for a better reading experience.

2. **Increase our external visibility**. A nicely formatted and readable public website would get more viewership than a plain repo, largely for the reasons stated in #1. External visibility is a good thing for several reasons:
    - *Reinforce our commitment to an open culture and OSS*
    - *Community involvement*. It's easier to solicit thoughtful feedback from the community on the patterns we've landed on, and surface any problems we may have overlooked, if our guides are highly visible
    - *Recruitment*. All else equal, companies with more OSS visibility tend to have an easier time recruiting

3. **Increase in internal use**. Projects with higher external visibility tend to be better maintained and receive more contributions from the team. As an anecodote, if we make the Tech blog public (and beautiful), I would write on it much more than I do now. Also, speaking for the Ember devs, our more popular Ember work gets more attention and love than our less visible work.

**Why use Ember?**

Those are my arguments for why `code-style-guides` should be a website. Here's why I think that website should be Made with Ember, in particular as opposed to Jekyll/Middleman:

1. **Ease of adding dynamic UI elements**. I've worked on several docs sites in OSS-land and inevitably, some dynamic JS behavior will be desired. Using Ember means we have access to the entire addon ecosystem (dropdowns, collapsibles, etc).

2. **Ease of nontrivial routing**. Adding a nontrivial routing schema (for example, to support versioned docs) tends to be easier in Ember than a static site generator. SSGs mainly use the directory structure itself as their routing mechanism, which is both a blessing and a curse. New concepts are added (e.g. Jekyll Collections) to give devs more power over their site structure, which ultimately negates some of the benefits of using a SSG in the first place.

3. **We use Ember at TED**. This means
    - *We can leverage shared solutions*. Things we know and learn in one Ember app can be used in another, including this one. Configuring Jenkins to build + deploy this app will be easier than a SSG, since we already know how to configure Jenkins 
    - *Non-Ember devs have another chance to learn Ember**. Many non-Ember devs at TED have expressed interest in learning Ember. If in the course of making updates to `code-style-guides`, a non-Ember dev picks up 5% of Ember's surface area, I see this as better than that dev learning 30% of Jekyll (or some other SSG).
        + I want to clarify that, in the normal course of things, I am confident that everyone on our team will be able to edit the guides without having to know anything about Ember, the same way many OSS contributors have been able to edit Mirage's guides without having to know anything about Jekyll.

4. **You can execute code in your style guide**. In the event you want to write a guide that's *more* than just a Markdown file, you can.

---

Finally, I'll touch on what I saw as Joe's main concern, which was that the repo itself is no longer simple Markdown documents and is instead a full-fledged Ember app. His concern was that this led to the actual Markdown content being hidden deep down within Ember's structure, making them hard to find for viewing and editing.

My response is that we will be able to structure the website in such a way that users will primarily edit the Markdown guides via "Improve this guide" links, similar to what I've done on Mirage's site [example](http://www.ember-cli-mirage.com/docs/v0.1.x/). Anecdotally, when I added this link to each guide, I saw a definite increase in contribution. If our docs site is clean + has a simple way to get users to a PR of each particular guide, in my mind this issue would be resolved.

Futher, because the Broccoli plugin allows us to continue to author in pure Markdown, the guides stay far more portable than a traditional Ember app comprised of HTMLbars templates.

---

Ultimately, I believe strongly in the benefits of an active and healthy `code-style-guides` for our growing team. The repo has been around for six months and has seen very little activity. Making it into an Ember app is my best attempt at a solution. If we end up nixing the Ember app, I hope we can arrive at an alternative workflow that would actually lead to this content/repo being used.

Ok, I'd love to hear everyone's thoughts on this!

**Comment by samselikoff** [comment]
/cc @redoPop @ryanto @brenna 

Just wanted to @ you all to let you know, there's no rush on this discussion.

**Comment by ryanto** [comment]
A few thoughts.

* My first reaction was that pod structure + markdown files would be hard to follow. After clicking around the pods folder it's pretty easy to understand how things work. A concern, you'd still need to know Ember to add a new page? That seems like it could be problematic.
* Can we get best of both worlds? Flat markdown, but transpile into HTML + clickable ember app. That way you can add with markdown and share either markdown or HTML.
* Since its a Broccoli plugin we'd need deployment. This makes it harder to make changes live. Also would require node + ember setup.
* There's something about style guides as HTML (with layout, links, etc) that make me take them a bit more seriously... I think it shows thought and dedication.
* Being an Ember app might cause an artificial barrier to entry. This might not be true, but it's something to consider.

---

Are there any issues by doing both markdown and html? If we can make the process as easy as adding a flat markdown file?

**Comment by brenna** [comment]
I've taken so long to respond to this because I'm essentially torn between the two options.

I can see both an accessible file structure (i.e flat markdown, sans Ember) and a polished HTML docs site being beneficial for this repo.  The former makes it easier to contribute, the latter encourages thoughtful contribution.

I like where Ryan's going with the idea of transpiling into an HTML docs site. There would be some setup overhead but if we could automate updates and deployment, we'd get the best of both worlds. 

Is Jekyll any closer to that middle ground or is the file structure still a barrier? At least with that we get easy deploys via gh-pages.

**Comment by redoPop** [comment]
+1 for compiling [tedconf.github.io/code-style-guides](https://tedconf.github.io/code-style-guides/) from flat markdown, with one `.md` file for each domain of concern.

I completely agree with Sam about the benefits of having this site. To me the question raised in this issue is not whether such a site should exist, or even whether it should be an Ember app or something else, but whether it's acceptable for our code style guides to be wrapped in the presentational code used to produce a site. My ideal alternative is a site generated from flat markdown sources, but until we've produced or decided on this best of both worlds platform I believe the benefits of having flat markdown source files outweigh the benefits of having a site.

Ryan, Sam, and Brenna have all touched on the approachability aspect, and Sam's done a lot of work to mitigate approachability issues by creating a system that uses markdown within pods, but even with this effort, storing presentation and content together still seems problematic to me.

Markdown's killer feature is that it is especially _human readable_ markup, and burying it in an Ember app diminishes that readability. As of b3a2875dfe6a19beaf74ee5f3763ac1e11fe7f6d, none of the repo's top level files actually contain code style guides: those markdown sources are buried in a tree three subdirectories deep, and within those subdirectories markdown is split further into multiple additional subdirectories matching the Ember app's routes. This presents much more friction for would-be contributors than the original flat file layout.

Training users how to find and add style guide sources in this layout may be regarded as a matter of documentation, but the problem won't go away after a single training: when our Ember style guides are updated to recommend a layout other than pods, markdown sources will need to be moved again and users will need to re-train their habits to work with the new layout.

For users who actively work with Ember, this might be somewhat felicitous: as Sam points out, it's another opportunity to learn the framework and our best practices. But I would prefer that style guide contributions be made as open and as low-friction as possible for everyone, even for people who lack familiarity with Ember.

Another aspect of approachability is the style guides' readability for offline use. Assuming I have the code style guides repository checked out locally, I can easily read and even `⌘F` flat markdown files in any text editor. In the current repository, I would instead need to do a directory search constrained to the `app/pods` subdirectory I'm interested in. This too adds friction and reduces approachability.

Finally, git commits and their linked PR conversations will be extremely helpful artifacts to examine when revisiting & reassessing style guide recommendations, and intertwining documentation with presentational code pollutes that history. Git is generally pretty smart about preserving history when whole files are moved intact, but splitting markdown apart and subjecting it to the layout and refactoring needs of a surrounding Ember app increases the amount of history layered in between true style guide changes and will make unwinding the history of a particular recommendation a more fricative process.

**Comment by samselikoff** [comment]
Good discush y'all. I just wanted to reiterate a few of my thoughts.

I personally don't see storing the Markdown guides alongside presentation code as a problem, because a well-designed site with good links will be how users discover the guides + how they get to editing them. Of course this is a hypothesis, but my experience with Mirage docs validates this, and also if you think about how you edit sites like Wikipedia, it seems to hold.

The readability of Markdown is important, but primarily at the point when users are already making edits. If we make a site, but users still use GH to *navigate and consume* the documentation, then the site has failed plain and simple, and imo it shouldn't exist.

Basically, my argument is that a decently structured site should supplant GitHub as a discovery mechanism for viewing *and* editing the guides (where editing is, a link to the `/edit` version of the appropriate doc on GH).

This also applies to Joe's concern about training people to find docs within the sourcecode. This was never my intention. The goal would be to have it behave just like Wikipedia: well-structured docs site solves the discoverability problem, a simple link to "edit" solves the problem of finding the appropriate file in the sourcecode.

Adding a new file is the exception here.

The point about having the simple `md` files offline is a good one, definitely a con of the Website approach.

Again, overall I'm more interested in style-guides being effective for our team than any particular solution. If we don't make style-guides a Website - i.e., if we don't change anything - what are we going to do to prevent another 6 months going by with 2-3 changes and practically no use within our team?

**Comment by samselikoff** [comment]
I'm reminded here of the build-measure-learn cycle from Lean Startup. We've spent a lot of time even just here writing + discussing + thinking - can we think of a way to run an experiment + let the team and us (our customers) give us feedback so we can learn?

**Comment by redoPop** [comment]
I love the Edit button idea.  Familiarity with Ember's routing and layout would still be needed to add and remove content sections – important use cases in this context – which means it does not mitigate the "approachability for editing" issue to my satisfaction. But it's still a great idea that deserves implementation in our style guides website.

> If we don't make style-guides a Website - i.e., if we don't change anything - what are we going to do to prevent another 6 months going by with 2-3 changes and practically no use within our team?

I'm a bit wary of the phrase "If we don't make style-guides a Website": per Ryan & Brenna's comments, having flat markdown sources & having a website are not mutually exclusive ideas.

"What are we going to do…" is a much broader brainstorming question encompassing outreach like Brenna's presentation to the team earlier this month. As much as possible, I'd like to keep discussion in this particular GH issue focused on a flat markdown layout vs. an Ember app (or other presentationally-focused) repo layout.

> can we think of a way to run an experiment + let the team and us (our customers) give us feedback so we can learn?

[Team members who contributed to this repository](https://github.com/tedconf/code-style-guides/graphs/contributors) prior to b3a2875dfe6a19beaf74ee5f3763ac1e11fe7f6d may be able to offer some feedback/opinions about the new layout, but I believe it's premature to raise this specific issue with the entire team:

As a team, TED Tech has not yet established the problems we'd like to see addressed by TED's code style guides that are not yet [addressed by Rubucop and other style guides & tools](https://github.com/tedconf/code-style-guides/blob/2b869f17557a5452b3516d89b48cc4f05a0e4a70/ruby.md). Nor have we established any expectations about contributions to this guide.

I've been fretting about how to word this response because I think _this_ issue is an important one since it dictates a pattern that everyone will be expected to follow, but it also presupposes some other things about code style guides that need to be discussed with the team first. We may need to put this issue on ice if we'd like the whole team's perspective before making a decision.

**Comment by samselikoff** [comment]
I agree that this issue should be iced if we're not on the same page about what problem code-style-guides is solving.

My take is that we currently have no place to put and share common code styles, heuristics, conventions, patterns, etc. that go *beyond* jshint and rubocop (the kinds of things in the markdown files as of 2207b126a7074153e56a0aa681f2ffac0a45b17e), and that we need such a place, and that code-style-guides was that place. If we're not all in agreement here, I guess we need to discuss.

One more thought, perhaps we could make a separate place for Ember guides, so we can experiment there, and not have to make this a big thing that involves everybody?

One last thought, I'd prefer some tinkering to more extended discussion, I feel like we'll all agree on a solution once we *see* it and *feel* it. The Ember app was my tinkering bc it seemed code-style-guides was not being unused. Maybe I could just move it to a separate repo and mess around myself, perhaps keep it to Ember guides, and we could see how it feels, and show it to Dan/Grady etc.? That way we'll start getting actual feedback as quickly as possible, instead of theorizing about what will or will not work.

Okay bed time because I'm still in :camel:

**Comment by samselikoff** [comment]
8fec9b50544b1ca15a69a192737c01594e740a82 closes
